sky/global_user_state.py

```diff
@@ -255,12 +337,58 @@ def set_cluster_metadata(cluster_name: str, metadata: Dict[str, Any]) -> None:
         raise ValueError(f'Cluster {cluster_name} not found.')
 
 
+def _get_cluster_usage_intervals(cluster_hash: str) -> Optional[Dict[str, Any]]:
+    rows = _DB.cursor.execute(
+        'SELECT usage_intervals FROM cluster_history WHERE cluster_hash=(?)',
+        (cluster_hash,))
+    for (usage_intervals,) in rows:
+        if usage_intervals is None:
+            return None
+        return pickle.loads(usage_intervals)
+
+
+def _set_cluster_usage_intervals(cluster_hash: str,
+                                 usage_intervals: Dict[str, Any]) -> None:
+    _DB.cursor.execute(
+        'UPDATE cluster_history SET usage_intervals=(?) WHERE cluster_hash=(?)',
+        (
+            pickle.dumps(usage_intervals),
+            cluster_hash,
+        ))
+    count = _DB.cursor.rowcount
+    _DB.conn.commit()
+    assert count <= 1, count
+    if count == 0:
+        raise ValueError(f'Cluster hash {cluster_hash} not found.')
+
+
+def _get_cluster_hash(cluster_name: str) -> Optional[str]:
+    rows = _DB.cursor.execute(
+        'SELECT cluster_hash FROM clusters WHERE name=(?)', (cluster_name,))
+    for (cluster_hash,) in rows:
+        if cluster_hash is None:
+            return None
+        return cluster_hash
+
+
+def get_launched_resources_from_cluster_hash(
+        cluster_hash: str) -> Optional[Tuple[int, Any]]:
+    rows = _DB.cursor.execute(
+        'SELECT launched_resources FROM cluster_history WHERE cluster_hash=(?)',
```

Comment:
can we rename the `launched_resources`, as it is a bit confusing that here it means a tuple of `(num_nodes, launched_resources)`. How about we split it into two columns, `num_nodes` and `launched_resources`, so that the `requested_resources` can share the `num_nodes`?